In [884]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
from lightgbm import LGBMRegressor, LGBMClassifier
import sys
import re
from sklearn.preprocessing import LabelEncoder
import plotly.express as px
from scipy.stats import spearmanr

In [885]:
df_listings = pd.read_csv("../data/Listing/listed_transactions.csv")
df_listings.head()

C:\Users\mayab\AppData\Local\Temp\ipykernel_30128\3410094314.py:1: DtypeWarning: Columns (0: ListAgentEmail, 1: BuyerAgencyCompensationType) have mixed types. Specify dtype option on import or set low_memory=False.
  df_listings = pd.read_csv("../data/Listing/listed_transactions.csv")


,OriginalListPrice,ListingKey,ListAgentEmail,CloseDate,ClosePrice,ListAgentFirstName,ListAgentLastName,Latitude,Longitude,UnparsedAddress,PropertyType,LivingArea,ListPrice,DaysOnMarket,ListOfficeName,BuyerOfficeName,CoListOfficeName,ListAgentFullName,CoListAgentFirstName,CoListAgentLastName,BuyerAgentMlsId,BuyerAgentFirstName,BuyerAgentLastName,FireplacesTotal,AssociationFeeFrequency,AboveGradeFinishedArea,ListingKeyNumeric,MLSAreaMajor,TaxAnnualAmount,CountyOrParish,PropertyType.1,MlsStatus,ElementarySchool,ListAgentFirstName.1,AttachedGarageYN,ParkingTotal,BuilderName,PropertySubType,LotSizeAcres,SubdivisionName,BuyerOfficeAOR,YearBuilt,DaysOnMarket.1,BuyerAgencyCompensationType,StreetNumberNumeric,LivingArea.1,ListingId,BathroomsTotalInteger,City,BuyerAgencyCompensation,TaxYear,BuildingAreaTotal,BedroomsTotal,ContractStatusChangeDate,Longitude.1,ElementarySchoolDistrict,CoBuyerAgentFirstName,PurchaseContractDate,ListingContractDate,BelowGradeFinishedArea,BusinessType,Latitude.1,ListPrice.1,StateOrProvince,CoveredSpaces,MiddleOrJuniorSchool,FireplaceYN,Stories,HighSchool,Levels,ListAgentLastName.1,CloseDate.1,LotSizeDimensions,LotSizeArea,MainLevelBedrooms,NewConstructionYN,GarageSpaces,HighSchoolDistrict,PostalCode,BuyerOfficeName.1,AssociationFee,LotSizeSquareFeet,MiddleOrJuniorSchoolDistrict,UnparsedAddress.1,Year,year_month,rate_30yr_fixed
0,90000.0,1075010398,miriamlara03@gmail.com,NaN,NaN,Miriam,Lara,34.097939,-117.909653,1045 N Azusa 61,ManufacturedInPark,960.0,90000.0,0,CENTURY 21 CITRUS REALTY INC,NaN,NaN,Miriam Lara,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,1075010398,614 - Covina,NaN,Los Angeles,ManufacturedInPark,Active,NaN,Miriam,False,4.0,NaN,NaN,NaN,NaN,NaN,1960.0,0,Item,1045.0,960.0,CV24093347,1.0,Covina,2500.0,NaN,NaN,4.0,2024-01-29,-117.909653,NaN,NaN,NaN,2024-01-29,NaN,NaN,34.097939,90000.0,CA,NaN,NaN,NaN,1.0,NaN,One,Lara,NaN,NaN,NaN,NaN,NaN,2.0,Covina Valley Unified,91722,NaN,0.0,NaN,NaN,1045 N Azusa 61,2024,2024-01,6.6425
1,1500000.0,1074974457,janelle@judsonre.com,NaN,NaN,Janelle,Judson,33.121241,-117.081614,NaN,CommercialSale,NaN,1500000.0,0,Judson Real Estate,NaN,Judson Real Estate,Janelle Judson,Lynn,Judson,NaN,NaN,NaN,NaN,NaN,NaN,1074974457,92025 - Escondido,NaN,San Diego,CommercialSale,Active,NaN,Janelle,NaN,0.0,NaN,Retail,NaN,NaN,NaN,1940.0,0,Item1,NaN,NaN,NDP2403932,0.0,Escondido,2.0,NaN,3000.0,NaN,2024-01-11,-117.081614,NaN,NaN,NaN,2024-01-11,NaN,NaN,33.121241,1500000.0,CA,NaN,NaN,NaN,NaN,NaN,NaN,Judson,NaN,NaN,3052.0,NaN,NaN,NaN,NaN,92025,NaN,0.0,NaN,NaN,NaN,2024,2024-01,6.6425
2,1340000.0,1074973329,haleh360@Gmail.com,NaN,NaN,Haleh,Dowlatshahi,34.052207,-118.408445,2220 Avenue Of The Stars 2704,Residential,1301.0,1340000.0,127,Rodeo Realty- Brentwood,NaN,NaN,Haleh Dowlatshahi,NaN,NaN,NaN,NaN,NaN,NaN,Monthly,NaN,1074973329,C05 - Westwood - Century City,NaN,Los Angeles,Residential,Pending,NaN,Haleh,NaN,0.0,NaN,Condominium,4.0831,NaN,NaN,1964.0,127,Item1,2220.0,1301.0,24389291,2.0,Los Angeles,2.0,NaN,1301.0,2.0,2024-05-07,-118.408445,NaN,NaN,2024-05-07,2024-01-01,NaN,NaN,34.052207,1340000.0,CA,NaN,NaN,False,1.0,NaN,One,Dowlatshahi,NaN,NaN,177861.0,NaN,False,NaN,NaN,90067,NaN,2105.0,177861.0,NaN,2220 Avenue Of The Stars 2704,2024,2024-01,6.6425
3,2500000.0,1074954552,Reneechen@yourhomesoldguaranteed.com,NaN,NaN,Renee,Chen,33.496363,-117.691677,16 Palisades,Residential,2788.0,2500000.0,1,Your Home Sold Guaranteed Realty,NaN,NaN,Renee Chen,NaN,NaN,NaN,NaN,NaN,NaN,Monthly,NaN,1074954552,LNSLT - Salt Creek,NaN,Orange,Residential,Active,NaN,Renee,True,3.0,NaN,SingleFamilyResidence,0.1217,Warmington Homes (BB) (BBW),NaN,1994.0,1,Item1,16.0,2788.0,AR24091628,3.0,Laguna Niguel,2.0,NaN,NaN,4.0,2024-01-24,-117.691677,NaN,NaN,NaN,2024-01-24,NaN,NaN,33.496363,2500000.0,CA,NaN,NaN,True,2.0,NaN,Two,Chen,NaN,NaN,5300.0,0.0,False,3.0,Capistrano Unified,92677,NaN,254.0,5300.0,NaN,16 Palisades,2024,2024-01,6.6425
4,3150000.0,1074936537,anader@dppre.com,NaN,NaN,Margaret,Nader,34.119345,-118.111254,1615 Waverly Road,Residentia

##### Number of Rows and Columns

In [886]:
df_listings.shape

(853754, 87)

The listing transactions file has 853754 rows and 87 columns. 

##### Review column data types

In [887]:
df_listings.info()

<class 'pandas.DataFrame'>
RangeIndex: 853754 entries, 0 to 853753
Data columns (total 87 columns):
 #   Column                        Non-Null Count   Dtype  
---  ------                        --------------   -----  
 0   OriginalListPrice             850444 non-null  float64
 1   ListingKey                    853754 non-null  int64  
 2   ListAgentEmail                780379 non-null  str    
 3   CloseDate                     255893 non-null  str    
 4   ClosePrice                    233059 non-null  float64
 5   ListAgentFirstName            848740 non-null  str    
 6   ListAgentLastName             853678 non-null  str    
 7   Latitude                      741807 non-null  float64
 8   Longitude                     742511 non-null  float64
 9   UnparsedAddress               851597 non-null  str    
 10  PropertyType                  853754 non-null  str    
 11  LivingArea                    747280 non-null  float64
 12  ListPrice                     851619 non-null  float64


In [888]:
df_listings.columns

Index(['OriginalListPrice', 'ListingKey', 'ListAgentEmail', 'CloseDate',
       'ClosePrice', 'ListAgentFirstName', 'ListAgentLastName', 'Latitude',
       'Longitude', 'UnparsedAddress', 'PropertyType', 'LivingArea',
       'ListPrice', 'DaysOnMarket', 'ListOfficeName', 'BuyerOfficeName',
       'CoListOfficeName', 'ListAgentFullName', 'CoListAgentFirstName',
       'CoListAgentLastName', 'BuyerAgentMlsId', 'BuyerAgentFirstName',
       'BuyerAgentLastName', 'FireplacesTotal', 'AssociationFeeFrequency',
       'AboveGradeFinishedArea', 'ListingKeyNumeric', 'MLSAreaMajor',
       'TaxAnnualAmount', 'CountyOrParish', 'PropertyType.1', 'MlsStatus',
       'ElementarySchool', 'ListAgentFirstName.1', 'AttachedGarageYN',
       'ParkingTotal', 'BuilderName', 'PropertySubType', 'LotSizeAcres',
       'SubdivisionName', 'BuyerOfficeAOR', 'YearBuilt', 'DaysOnMarket.1',
       'BuyerAgencyCompensationType', 'StreetNumberNumeric', 'LivingArea.1',
       'ListingId', 'BathroomsTotalInteger', 'Cit

In [889]:
df_listings.isna().sum()

OriginalListPrice                 3310
ListingKey                           0
ListAgentEmail                   73375
CloseDate                       597861
ClosePrice                      620695
ListAgentFirstName                5014
ListAgentLastName                   76
Latitude                        111947
Longitude                       111243
UnparsedAddress                   2157
PropertyType                         0
LivingArea                      106474
ListPrice                         2135
DaysOnMarket                         0
ListOfficeName                       0
BuyerOfficeName                 608260
CoListOfficeName                682807
ListAgentFullName                  228
CoListAgentFirstName            683440
CoListAgentLastName             682952
BuyerAgentMlsId                 606667
BuyerAgentFirstName             606829
BuyerAgentLastName              605919
FireplacesTotal                 853754
AssociationFeeFrequency         607361
AboveGradeFinishedArea   

#### **Missing Value Analysis**

##### Calculate missing counts and percentages per column

In [890]:
missing_listing_counts = df_listings.isnull().sum()
missing_listing_counts

OriginalListPrice                 3310
ListingKey                           0
ListAgentEmail                   73375
CloseDate                       597861
ClosePrice                      620695
ListAgentFirstName                5014
ListAgentLastName                   76
Latitude                        111947
Longitude                       111243
UnparsedAddress                   2157
PropertyType                         0
LivingArea                      106474
ListPrice                         2135
DaysOnMarket                         0
ListOfficeName                       0
BuyerOfficeName                 608260
CoListOfficeName                682807
ListAgentFullName                  228
CoListAgentFirstName            683440
CoListAgentLastName             682952
BuyerAgentMlsId                 606667
BuyerAgentFirstName             606829
BuyerAgentLastName              605919
FireplacesTotal                 853754
AssociationFeeFrequency         607361
AboveGradeFinishedArea   

In [891]:
missing_listing_percent = (df_listings.isnull().mean()) * 100
missing_listing_percent

OriginalListPrice                 0.387700
ListingKey                        0.000000
ListAgentEmail                    8.594396
CloseDate                        70.027315
ClosePrice                       72.701856
ListAgentFirstName                0.587289
ListAgentLastName                 0.008902
Latitude                         13.112325
Longitude                        13.029866
UnparsedAddress                   0.252649
PropertyType                      0.000000
LivingArea                       12.471274
ListPrice                         0.250072
DaysOnMarket                      0.000000
ListOfficeName                    0.000000
BuyerOfficeName                  71.245347
CoListOfficeName                 79.977019
ListAgentFullName                 0.026706
CoListAgentFirstName             80.051162
CoListAgentLastName              79.994003
BuyerAgentMlsId                  71.058759
BuyerAgentFirstName              71.077734
BuyerAgentLastName               70.971146
FireplacesT

In [892]:
missing_summary = pd.DataFrame({
    "missing_listing_counts": missing_listing_counts,
    "missing_listing_percent": missing_listing_percent
})

In [893]:
missing_listing_summary = missing_summary.sort_values(by="missing_listing_percent", ascending=False)
print(missing_listing_summary)

                              missing_listing_counts  missing_listing_percent
CoveredSpaces                                 853754               100.000000
AboveGradeFinishedArea                        853754               100.000000
MiddleOrJuniorSchoolDistrict                  853754               100.000000
ElementarySchoolDistrict                      853754               100.000000
FireplacesTotal                               853754               100.000000
TaxYear                                       852886                99.898331
BelowGradeFinishedArea                        850334                99.599416
BusinessType                                  847779                99.300150
TaxAnnualAmount                               847661                99.286328
CoBuyerAgentFirstName                         834049                97.691958
BuilderName                                   824452                96.567864
LotSizeDimensions                             805053            

In [894]:
missing_listing_summary = missing_listing_summary[missing_listing_summary["missing_listing_counts"] > 0]
print(missing_listing_summary)

                              missing_listing_counts  missing_listing_percent
CoveredSpaces                                 853754               100.000000
AboveGradeFinishedArea                        853754               100.000000
MiddleOrJuniorSchoolDistrict                  853754               100.000000
ElementarySchoolDistrict                      853754               100.000000
FireplacesTotal                               853754               100.000000
TaxYear                                       852886                99.898331
BelowGradeFinishedArea                        850334                99.599416
BusinessType                                  847779                99.300150
TaxAnnualAmount                               847661                99.286328
CoBuyerAgentFirstName                         834049                97.691958
BuilderName                                   824452                96.567864
LotSizeDimensions                             805053            

In [895]:
missing_listing_summary["missing_listing_percent"] = missing_listing_summary["missing_listing_percent"].round(2)
print(missing_listing_summary)

                              missing_listing_counts  missing_listing_percent
CoveredSpaces                                 853754                   100.00
AboveGradeFinishedArea                        853754                   100.00
MiddleOrJuniorSchoolDistrict                  853754                   100.00
ElementarySchoolDistrict                      853754                   100.00
FireplacesTotal                               853754                   100.00
TaxYear                                       852886                    99.90
BelowGradeFinishedArea                        850334                    99.60
BusinessType                                  847779                    99.30
TaxAnnualAmount                               847661                    99.29
CoBuyerAgentFirstName                         834049                    97.69
BuilderName                                   824452                    96.57
LotSizeDimensions                             805053            

In [896]:
missing_above_90 = missing_listing_summary[missing_listing_summary['missing_listing_percent'] > 90]
print(missing_above_90)

                              missing_listing_counts  missing_listing_percent
CoveredSpaces                                 853754                   100.00
AboveGradeFinishedArea                        853754                   100.00
MiddleOrJuniorSchoolDistrict                  853754                   100.00
ElementarySchoolDistrict                      853754                   100.00
FireplacesTotal                               853754                   100.00
TaxYear                                       852886                    99.90
BelowGradeFinishedArea                        850334                    99.60
BusinessType                                  847779                    99.30
TaxAnnualAmount                               847661                    99.29
CoBuyerAgentFirstName                         834049                    97.69
BuilderName                                   824452                    96.57
LotSizeDimensions                             805053            

In [897]:
missing_above_70 = missing_listing_summary[missing_listing_summary['missing_listing_percent'] > 70]
print(missing_above_70)

                              missing_listing_counts  missing_listing_percent
CoveredSpaces                                 853754                   100.00
AboveGradeFinishedArea                        853754                   100.00
MiddleOrJuniorSchoolDistrict                  853754                   100.00
ElementarySchoolDistrict                      853754                   100.00
FireplacesTotal                               853754                   100.00
TaxYear                                       852886                    99.90
BelowGradeFinishedArea                        850334                    99.60
BusinessType                                  847779                    99.30
TaxAnnualAmount                               847661                    99.29
CoBuyerAgentFirstName                         834049                    97.69
BuilderName                                   824452                    96.57
LotSizeDimensions                             805053            

In [898]:
missing_above_70.shape

(31, 2)

In [899]:
missing_above_70

,missing_listing_counts,missing_listing_percent
CoveredSpaces,853754,100.00
AboveGradeFinishedArea,853754,100.00
MiddleOrJuniorSchoolDistrict,853754,100.00
ElementarySchoolDistrict,853754,100.00
FireplacesTotal,853754,100.00
TaxYear,852886,99.90
BelowGradeFinishedArea,850334,99.60
BusinessType,847779,99.30
TaxAnnualAmount,847661,99.29
CoBuyerAgentFirstName,834049,97.69


To perserve the variables, it would be sufficient if each variable contain less than 70% of missing values in order to use imputation to replace the missingness in the data. 

So, variables with 70% of missing variables will be removed from the dataset because through imputation with variables with a significant number of missingness, it may contribute to bias towards certain values than others. 

Keep these key variables (from above 70):
- ClosePrice
- CloseDate

In [900]:
df_listings_clean = df_listings.drop(columns=['MiddleOrJuniorSchoolDistrict', 'ElementarySchoolDistrict', 'CoveredSpaces',
                                             'AboveGradeFinishedArea', 'FireplacesTotal', 'TaxYear', 
                                             'BelowGradeFinishedArea', 'BusinessType', 'TaxAnnualAmount',
                                             'CoBuyerAgentFirstName', 'BuilderName', 'LotSizeDimensions',
                                             'ElementarySchool', 'MiddleOrJuniorSchool', 'HighSchool',
                                             'BuyerAgencyCompensation', 'BuyerAgencyCompensationType', 'BuildingAreaTotal',
                                             'CoListAgentFirstName', 'CoListAgentLastName', 'CoListOfficeName', 'BuyerOfficeAOR',
                                             'BuyerOfficeName', 'BuyerOfficeName.1', 'AssociationFeeFrequency', 'BuyerAgentFirstName',
                                             'BuyerAgentMlsId', 'BuyerAgentLastName', 'CloseDate.1'])

Dropping Variables:
- MiddleOrJuniorSchoolDistrict                  
- ElementarySchoolDistrict
- CoveredSpaces
- AboveGradeFinishedArea
- FireplacesTotal
- TaxYear
- BelowGradeFinishedArea
- BusinessType
- TaxAnnualAmount
- CoBuyerAgentFirstName
- BuilderName
- LotSizeDimensions
- ElementarySchool
- MiddleOrJuniorSchool

In [901]:
# Calculate percentage for all columns
missing_pct = (df_listings_clean.isna().sum() / len(df_listings_clean)) * 100

# Filter to show only columns that actually have missing data
print(missing_pct[missing_pct > 0].sort_values(ascending=False))

ClosePrice                  72.701856
CloseDate                   70.027315
SubdivisionName             66.606306
PurchaseContractDate        56.868372
MainLevelBedrooms           53.761622
HighSchoolDistrict          39.757237
AttachedGarageYN            34.093193
AssociationFee              32.320200
Stories                     25.941899
Levels                      19.215371
GarageSpaces                18.987085
FireplaceYN                 14.354369
NewConstructionYN           13.437243
Latitude                    13.112325
Latitude.1                  13.112325
Longitude.1                 13.029866
Longitude                   13.029866
LivingArea                  12.471274
LivingArea.1                12.471274
PropertySubType             12.201407
BedroomsTotal               12.073853
MLSAreaMajor                11.293066
LotSizeAcres                 9.554392
LotSizeSquareFeet            9.237556
LotSizeArea                  9.111875
ListAgentEmail               8.594396
BathroomsTot

In [902]:
num_cols_with_missing = df_listings_clean.isna().any().sum()

print(f"Number of variables with missing values: {num_cols_with_missing}")

Number of variables with missing values: 45


In [903]:
cols_rows_dropping = missing_pct[(missing_pct < 1) & (missing_pct > 0)]
cols_rows_dropping

OriginalListPrice           0.387700
ListAgentFirstName          0.587289
ListAgentLastName           0.008902
UnparsedAddress             0.252649
ListPrice                   0.250072
ListAgentFullName           0.026706
CountyOrParish              0.000117
ListAgentFirstName.1        0.587289
StreetNumberNumeric         0.441813
City                        0.112562
ContractStatusChangeDate    0.809484
ListPrice.1                 0.250072
StateOrProvince             0.007731
ListAgentLastName.1         0.008902
PostalCode                  0.024363
UnparsedAddress.1           0.252649
dtype: float64

In [904]:
cols_rows_dropped = ['OriginalListPrice', 'ListAgentFirstName', 'ListAgentLastName', 'UnparsedAddress', 'ListPrice', 'ListAgentFullName', 'CountyOrParish',
        'StreetNumberNumeric', 'City', 'ContractStatusChangeDate','StateOrProvince', 'PostalCode']

In [905]:
df_listings_clean = df_listings_clean.dropna(subset=['OriginalListPrice', 'ListAgentFirstName', 'ListAgentLastName', 'UnparsedAddress', 'ListPrice', 'ListAgentFullName', 'CountyOrParish',
        'StreetNumberNumeric', 'City', 'ContractStatusChangeDate','StateOrProvince', 'PostalCode'])

In [906]:
df_listings_clean.shape

(833710, 58)

In [907]:
# Create a reference mapping of Names -> Emails
# Drop any rows that have missing emails so the map is 'clean'
email_lookup = df_listings_clean.dropna(subset=['ListAgentEmail']).drop_duplicates(['ListAgentFirstName', 'ListAgentLastName'])

# Set the names as the index to make the 'lookup' possible
email_lookup = email_lookup.set_index(['ListAgentFirstName', 'ListAgentLastName'])['ListAgentEmail']

# Create a function to apply the logic
def fill_missing_emails(row):
    # Only try to fill if the current email is missing
    if pd.isna(row['ListAgentEmail']):
        # Look up the name in our reference map
        return email_lookup.get((row['ListAgentFirstName'], row['ListAgentLastName']), "None")
    return row['ListAgentEmail']

# Apply it to the dataframe
df_listings_clean['ListAgentEmail'] = df_listings_clean.apply(fill_missing_emails, axis=1)

In [908]:
df_listings_clean['ListAgentEmail'].isna().sum()

np.int64(0)

In [909]:
(df_listings_clean['ListAgentEmail'] == 'None').sum()

np.int64(3297)

From the code above, I noticed that there are no missing values for list agent's first and last names, so to determine their emails, I tried to look up any matches by full name to get the list agent's emails for the rows that are missing them. If there are no matches, those emails will be preserved by having a value of "None" (i.e. no email provided).

In [910]:
df_listings_clean.isna().sum()

OriginalListPrice                0
ListingKey                       0
ListAgentEmail                   0
CloseDate                   582499
ClosePrice                  603953
ListAgentFirstName               0
ListAgentLastName                0
Latitude                    108404
Longitude                   107727
UnparsedAddress                  0
PropertyType                     0
LivingArea                  100952
ListPrice                        0
DaysOnMarket                     0
ListOfficeName                   0
ListAgentFullName                0
ListingKeyNumeric                0
MLSAreaMajor                 93897
CountyOrParish                   0
PropertyType.1                   0
MlsStatus                        0
ListAgentFirstName.1             0
AttachedGarageYN            283027
ParkingTotal                 49684
PropertySubType             100629
LotSizeAcres                 79474
SubdivisionName             555061
YearBuilt                    66207
DaysOnMarket.1      

In [911]:
df_listings_clean.columns

Index(['OriginalListPrice', 'ListingKey', 'ListAgentEmail', 'CloseDate',
       'ClosePrice', 'ListAgentFirstName', 'ListAgentLastName', 'Latitude',
       'Longitude', 'UnparsedAddress', 'PropertyType', 'LivingArea',
       'ListPrice', 'DaysOnMarket', 'ListOfficeName', 'ListAgentFullName',
       'ListingKeyNumeric', 'MLSAreaMajor', 'CountyOrParish', 'PropertyType.1',
       'MlsStatus', 'ListAgentFirstName.1', 'AttachedGarageYN', 'ParkingTotal',
       'PropertySubType', 'LotSizeAcres', 'SubdivisionName', 'YearBuilt',
       'DaysOnMarket.1', 'StreetNumberNumeric', 'LivingArea.1', 'ListingId',
       'BathroomsTotalInteger', 'City', 'BedroomsTotal',
       'ContractStatusChangeDate', 'Longitude.1', 'PurchaseContractDate',
       'ListingContractDate', 'Latitude.1', 'ListPrice.1', 'StateOrProvince',
       'FireplaceYN', 'Stories', 'Levels', 'ListAgentLastName.1',
       'LotSizeArea', 'MainLevelBedrooms', 'NewConstructionYN', 'GarageSpaces',
       'HighSchoolDistrict', 'PostalCode'

#### **Convert date fields to datetime format (CloseDate, PurchaseContractDate, ListingContractDate, ContractStatusChangeDate)**

In [912]:
df_listings_clean['CloseDate'] = pd.to_datetime(df_listings_clean['CloseDate'], errors='coerce')

In [913]:
df_listings_clean['PurchaseContractDate'] = pd.to_datetime(df_listings_clean['PurchaseContractDate'], errors='coerce')

In [914]:
df_listings_clean['ListingContractDate'] = pd.to_datetime(df_listings_clean['ListingContractDate'], errors='coerce')

In [915]:
df_listings_clean['ContractStatusChangeDate'] = pd.to_datetime(df_listings_clean['ContractStatusChangeDate'], errors='coerce')

In [916]:
print(df_listings_clean['PurchaseContractDate'].dtype)

datetime64[us]


#### **Remove unnecessary or redundant columns**

In [917]:
df_listings_clean.columns

Index(['OriginalListPrice', 'ListingKey', 'ListAgentEmail', 'CloseDate',
       'ClosePrice', 'ListAgentFirstName', 'ListAgentLastName', 'Latitude',
       'Longitude', 'UnparsedAddress', 'PropertyType', 'LivingArea',
       'ListPrice', 'DaysOnMarket', 'ListOfficeName', 'ListAgentFullName',
       'ListingKeyNumeric', 'MLSAreaMajor', 'CountyOrParish', 'PropertyType.1',
       'MlsStatus', 'ListAgentFirstName.1', 'AttachedGarageYN', 'ParkingTotal',
       'PropertySubType', 'LotSizeAcres', 'SubdivisionName', 'YearBuilt',
       'DaysOnMarket.1', 'StreetNumberNumeric', 'LivingArea.1', 'ListingId',
       'BathroomsTotalInteger', 'City', 'BedroomsTotal',
       'ContractStatusChangeDate', 'Longitude.1', 'PurchaseContractDate',
       'ListingContractDate', 'Latitude.1', 'ListPrice.1', 'StateOrProvince',
       'FireplaceYN', 'Stories', 'Levels', 'ListAgentLastName.1',
       'LotSizeArea', 'MainLevelBedrooms', 'NewConstructionYN', 'GarageSpaces',
       'HighSchoolDistrict', 'PostalCode'

In [918]:
df_listings_clean.shape

(833710, 58)

In [919]:
df_listings_clean = df_listings_clean.drop(columns=['PropertyType.1', 'ListAgentFirstName.1', 'DaysOnMarket.1', 'LivingArea.1', 'Longitude.1', 
                                                        'Latitude.1', 'ListPrice.1', 'ListAgentLastName.1', 'UnparsedAddress.1'])

In [920]:
df_listings_clean.shape

(833710, 49)

In [921]:
(df_listings_clean['OriginalListPrice'] == df_listings_clean['ListPrice']).value_counts()

True     670462
False    163248
Name: count, dtype: int64

#### **Outlier Cleaning**

Remove or flag invalid numeric values: ClosePrice <=0, LivingArea <=0, DaysOnMarket < 0, negative Bedrooms or Bathrooms

Removing and inspecting negative values

In [922]:
neg_counts = (df_listings_clean.select_dtypes(include='number') < 0).sum()

In [923]:
neg_counts[neg_counts > 0]

Latitude             7
Longitude       725767
DaysOnMarket        37
ParkingTotal       165
dtype: int64

In [924]:
(neg_counts[neg_counts > 0] / df_listings_clean.shape[0] * 100).sort_values(ascending=False)

Longitude       87.052692
ParkingTotal     0.019791
DaysOnMarket     0.004438
Latitude         0.000840
dtype: float64

ParkingTotal        

In [925]:
neg_ParkingTotal = df_listings_clean[df_listings_clean['ParkingTotal'] < 0]
neg_ParkingTotal.head()

,OriginalListPrice,ListingKey,ListAgentEmail,CloseDate,ClosePrice,ListAgentFirstName,ListAgentLastName,Latitude,Longitude,UnparsedAddress,PropertyType,LivingArea,ListPrice,DaysOnMarket,ListOfficeName,ListAgentFullName,ListingKeyNumeric,MLSAreaMajor,CountyOrParish,MlsStatus,AttachedGarageYN,ParkingTotal,PropertySubType,LotSizeAcres,SubdivisionName,YearBuilt,StreetNumberNumeric,ListingId,BathroomsTotalInteger,City,BedroomsTotal,ContractStatusChangeDate,PurchaseContractDate,ListingContractDate,StateOrProvince,FireplaceYN,Stories,Levels,LotSizeArea,MainLevelBedrooms,NewConstructionYN,GarageSpaces,HighSchoolDistrict,PostalCode,AssociationFee,LotSizeSquareFeet,Year,year_month,rate_30yr_fixed
8443,2300000.0,1058688466,mabroukhamza093@gmail.com,NaT,NaN,Hamza,Mabrouk,NaN,NaN,725 Hyde Street,CommercialSale,NaN,2300000.0,80,CENTURY 21 Real Estate Alliance,Hamza Mabrouk,1058688466,NaN,San Francisco,Active,NaN,-5.0,NaN,0.0379,NaN,1925.0,725.0,ML81951327,0.0,San Francisco,NaN,2024-01-24,NaT,2024-01-24,CA,NaN,NaN,NaN,1650.0,NaN,NaN,NaN,San Francisco Unified,94109,0.0,1650.0,2024,2024-01,6.6425
12223,1258000.0,1058416008,trisha@trishamotter.com,2024-02-15,1600000.0,Trisha,Motter,NaN,NaN,7385 Forsum Road,Residential,1693.0,1258000.0,6,E3 Realty & Loans,Trisha Motter,1058416008,699 - Not Defined,Santa Clara,Closed,True,-3.0,SingleFamilyResidence,0.1200,NaN,1992.0,7385.0,ML81950358,3.0,San Jose,4.0,2024-02-15,2024-01-25,2024-01-19,CA,True,NaN,NaN,5227.0,NaN,False,2.0,Other,95138,NaN,5227.0,2024,2024-01,6.6425
26306,1588888.0,1054056032,maryoproperties@yahoo.com,2024-02-23,1550000.0,Mary,O'neill,NaN,NaN,2965 Calle De Las Estrella,Residential,2578.0,1588888.0,22,Stevensen & Neal Realtors,Mary O'neill,1054056032,699 - Not Defined,Santa Clara,Closed,True,-1.0,SingleFamilyResidence,0.0900,NaN,1990.0,2965.0,ML81950466,3.0,San Jose,4.0,2024-02-23,2024-01-24,2024-01-02,CA,True,NaN,NaN,3920.0,NaN,False,2.0,Other,95148,120.0,3920.0,2024,2024-01,6.6425
30751,1599000.0,1061806243,kroyer.re@gmail.com,2024-04-02,1620000.0,Kathy,Royer,NaN,NaN,6088 Pietz Court,Residential,2184.0,1599000.0,5,Coldwell Banker Realty,Kathy Royer,1061806243,699 - Not Defined,Santa Clara,Closed,True,-139.0,SingleFamilyResidence,0.1400,NaN,1968.0,6088.0,ML81954879,3.0,San Jose,5.0,2024-04-02,2024-03-04,2024-02-28,CA,True,NaN,NaN,6098.0,NaN,False,2.0,San Jose Unified,95123,NaN,6098.0,2024,2024-02,6.7760
31804,1400000.0,1061770369,rick.talmage@compass.com,2024-03-27,1550000.0,Rick,Talmage,NaN,NaN,994 La Mesa Terrace E,Residential,2120.0,1400000.0,7,Compass,Rick Talmage,1061770369,699 - Not Defined,Santa Clara,Closed,True,-2.0,Townhouse,NaN,NaN,1985.0,994.0,ML81955654,3.0,Sunnyvale,2.0,2024-03-27,2024-03-06,2024-02-28,CA,True,NaN,NaN,NaN,NaN,False,2.0,Other,94086,655.0,NaN,2024,2024-02,6.7760


In [926]:
neg_ParkingTotal.head()

,OriginalListPrice,ListingKey,ListAgentEmail,CloseDate,ClosePrice,ListAgentFirstName,ListAgentLastName,Latitude,Longitude,UnparsedAddress,PropertyType,LivingArea,ListPrice,DaysOnMarket,ListOfficeName,ListAgentFullName,ListingKeyNumeric,MLSAreaMajor,CountyOrParish,MlsStatus,AttachedGarageYN,ParkingTotal,PropertySubType,LotSizeAcres,SubdivisionName,YearBuilt,StreetNumberNumeric,ListingId,BathroomsTotalInteger,City,BedroomsTotal,ContractStatusChangeDate,PurchaseContractDate,ListingContractDate,StateOrProvince,FireplaceYN,Stories,Levels,LotSizeArea,MainLevelBedrooms,NewConstructionYN,GarageSpaces,HighSchoolDistrict,PostalCode,AssociationFee,LotSizeSquareFeet,Year,year_month,rate_30yr_fixed
8443,2300000.0,1058688466,mabroukhamza093@gmail.com,NaT,NaN,Hamza,Mabrouk,NaN,NaN,725 Hyde Street,CommercialSale,NaN,2300000.0,80,CENTURY 21 Real Estate Alliance,Hamza Mabrouk,1058688466,NaN,San Francisco,Active,NaN,-5.0,NaN,0.0379,NaN,1925.0,725.0,ML81951327,0.0,San Francisco,NaN,2024-01-24,NaT,2024-01-24,CA,NaN,NaN,NaN,1650.0,NaN,NaN,NaN,San Francisco Unified,94109,0.0,1650.0,2024,2024-01,6.6425
12223,1258000.0,1058416008,trisha@trishamotter.com,2024-02-15,1600000.0,Trisha,Motter,NaN,NaN,7385 Forsum Road,Residential,1693.0,1258000.0,6,E3 Realty & Loans,Trisha Motter,1058416008,699 - Not Defined,Santa Clara,Closed,True,-3.0,SingleFamilyResidence,0.1200,NaN,1992.0,7385.0,ML81950358,3.0,San Jose,4.0,2024-02-15,2024-01-25,2024-01-19,CA,True,NaN,NaN,5227.0,NaN,False,2.0,Other,95138,NaN,5227.0,2024,2024-01,6.6425
26306,1588888.0,1054056032,maryoproperties@yahoo.com,2024-02-23,1550000.0,Mary,O'neill,NaN,NaN,2965 Calle De Las Estrella,Residential,2578.0,1588888.0,22,Stevensen & Neal Realtors,Mary O'neill,1054056032,699 - Not Defined,Santa Clara,Closed,True,-1.0,SingleFamilyResidence,0.0900,NaN,1990.0,2965.0,ML81950466,3.0,San Jose,4.0,2024-02-23,2024-01-24,2024-01-02,CA,True,NaN,NaN,3920.0,NaN,False,2.0,Other,95148,120.0,3920.0,2024,2024-01,6.6425
30751,1599000.0,1061806243,kroyer.re@gmail.com,2024-04-02,1620000.0,Kathy,Royer,NaN,NaN,6088 Pietz Court,Residential,2184.0,1599000.0,5,Coldwell Banker Realty,Kathy Royer,1061806243,699 - Not Defined,Santa Clara,Closed,True,-139.0,SingleFamilyResidence,0.1400,NaN,1968.0,6088.0,ML81954879,3.0,San Jose,5.0,2024-04-02,2024-03-04,2024-02-28,CA,True,NaN,NaN,6098.0,NaN,False,2.0,San Jose Unified,95123,NaN,6098.0,2024,2024-02,6.7760
31804,1400000.0,1061770369,rick.talmage@compass.com,2024-03-27,1550000.0,Rick,Talmage,NaN,NaN,994 La Mesa Terrace E,Residential,2120.0,1400000.0,7,Compass,Rick Talmage,1061770369,699 - Not Defined,Santa Clara,Closed,True,-2.0,Townhouse,NaN,NaN,1985.0,994.0,ML81955654,3.0,Sunnyvale,2.0,2024-03-27,2024-03-06,2024-02-28,CA,True,NaN,NaN,NaN,NaN,False,2.0,Other,94086,655.0,NaN,2024,2024-02,6.7760


In [927]:
df_listings_clean['ParkingTotal'].describe()

count    7.840260e+05
mean     6.505778e+00
std      2.938914e+03
min     -1.430000e+02
25%      1.000000e+00
50%      2.000000e+00
75%      3.000000e+00
max      2.593669e+06
Name: ParkingTotal, dtype: float64

DaysOnMarket              

In [928]:
neg_DaysOnMarket = df_listings_clean[df_listings_clean['DaysOnMarket'] < 0]
neg_DaysOnMarket.head()

,OriginalListPrice,ListingKey,ListAgentEmail,CloseDate,ClosePrice,ListAgentFirstName,ListAgentLastName,Latitude,Longitude,UnparsedAddress,PropertyType,LivingArea,ListPrice,DaysOnMarket,ListOfficeName,ListAgentFullName,ListingKeyNumeric,MLSAreaMajor,CountyOrParish,MlsStatus,AttachedGarageYN,ParkingTotal,PropertySubType,LotSizeAcres,SubdivisionName,YearBuilt,StreetNumberNumeric,ListingId,BathroomsTotalInteger,City,BedroomsTotal,ContractStatusChangeDate,PurchaseContractDate,ListingContractDate,StateOrProvince,FireplaceYN,Stories,Levels,LotSizeArea,MainLevelBedrooms,NewConstructionYN,GarageSpaces,HighSchoolDistrict,PostalCode,AssociationFee,LotSizeSquareFeet,Year,year_month,rate_30yr_fixed
223,799000.0,1063549350,ainsleyhughes@kw.com,2024-03-21,799000.0,Ainsley,Hughes,34.425577,-119.291855,11905 Silver Spur Street,Residential,1284.0,799000.0,-48,Keller Williams West Ventura County,Ainsley Hughes,1063549350,VC11 - Ojai,Ventura,Closed,True,2.0,SingleFamilyResidence,0.1351,NaN,1989.0,11905.0,V1-22556,2.0,Ojai,3.0,2024-03-21,2024-02-02,2024-01-30,CA,True,1.0,One,5885.0,NaN,NaN,2.0,NaN,93023,59.0,5885.0,2024,2024-01,6.6425
228,899000.0,1063528331,absea@comcast.net,2024-03-19,810000.0,A.B.,Priceman DRE 0126...,39.383806,-123.789254,31530 Emerald Drive,Residential,1164.0,899000.0,-58,Garden Ranch Real Estate,A.B. Priceman DRE 0126...,1063528331,FBS - Fort Bragg South,Mendocino,Closed,True,2.0,SingleFamilyResidence,0.8000,NaN,1974.0,31530.0,C1-10508,1.0,Fort Bragg,2.0,2024-03-19,2024-01-22,2024-01-11,CA,False,1.0,One,0.8,NaN,False,2.0,NaN,95437,NaN,34848.0,2024,2024-01,6.6425
752,11500.0,1061271257,teresa.fuller@compass.com,2024-03-07,10500.0,Teresa,Fuller,34.081125,-118.363403,536 N Edinburgh Avenue,ResidentialLease,2800.0,10500.0,-16,COMPASS,Teresa Fuller,1061271257,C10 - West Hollywood Vicinity,Los Angeles,Closed,NaN,0.0,SingleFamilyResidence,0.1518,Not Applicable,1924.0,536.0,P1-16535,4.0,Los Angeles,5.0,2024-03-07,2024-01-30,2024-01-30,CA,True,1.0,One,6612.0,5.0,NaN,0.0,NaN,90048,0.0,6612.0,2024,2024-01,6.6425
1119,1599000.0,1060153479,robert@anppros.com,2024-03-21,1625000.0,Robert,Perez,34.244449,-118.265167,3929 El Moreno Street,Residential,2130.0,1599000.0,-1,Evernest Real Estate Advisors,Robert Perez,1060153479,635 - La Crescenta/Glendale Montrose & Annex,Los Angeles,Closed,False,2.0,SingleFamilyResidence,0.1205,NaN,1964.0,3929.0,P1-16438,3.0,La Crescenta,4.0,2024-03-21,2024-02-06,2024-01-31,CA,True,2.0,Two,5249.0,NaN,NaN,2.0,NaN,91214,NaN,5249.0,2024,2024-01,6.6425
4167,469999.0,1059512539,khoren9@yahoo.com,NaT,NaN,Khoren,Barutyan,34.201700,-118.460053,15015 Sherman Way 103,Residential,1128.0,469999.0,-33,"KB Estates, Inc.",Khoren Barutyan,1059512539,VN - Van Nuys,Los Angeles,Active,False,2.0,Condominium,0.3575,NaN,1986.0,15015.0,SR24015836,3.0,Van Nuys,2.0,2024-01-25,NaT,2024-01-25,CA,False,1.0,"One,Two",15574.0,0.0,False,2.0,Los Angeles Unified,91405,390.0,15574.0,2024,2024-01,6.6425


In [929]:
rows_DaysOnMarket = neg_DaysOnMarket['DaysOnMarket']
rows_DaysOnMarket.head()

223    -48
228    -58
752    -16
1119    -1
4167   -33
Name: DaysOnMarket, dtype: int64

In [930]:
non_neg = (df_listings_clean['DaysOnMarket'] >= 0) & (df_listings_clean['ParkingTotal'] >= 0)

In [931]:
df_listings_clean = df_listings_clean[non_neg].copy()

In [932]:
df_listings_clean.shape

(783825, 49)

In [933]:
neg_counts = (df_listings_clean.select_dtypes(include='number') < 0).sum()

In [934]:
neg_counts[neg_counts > 0]

Latitude          4
Longitude    680607
dtype: int64

In [935]:
df_listings_clean = df_listings_clean[
    (df_listings_clean['Latitude'] >= 32.5) & (df_listings_clean['Latitude'] <= 42.0) &
    (df_listings_clean['Longitude'] >= -124.5) & (df_listings_clean['Longitude'] <= -114.0)
]

In [936]:
df_listings_clean.shape

(679755, 49)

In [937]:
'''
df_map = df_listings_clean.sample(30000)

fig = px.scatter_mapbox(
    df_map,
    lat="Latitude",
    lon="Longitude",
    color="OriginalListPrice",
    size_max=15,
    zoom=5,
    title="California Property Value Distribution",
    color_continuous_scale=px.colors.sequential.Plasma,
    mapbox_style="open-street-map"
)

fig.show(renderer='browser')
'''

'\ndf_map = df_listings_clean.sample(30000)\n\nfig = px.scatter_mapbox(\n    df_map,\n    lat="Latitude",\n    lon="Longitude",\n    color="OriginalListPrice",\n    size_max=15,\n    zoom=5,\n    title="California Property Value Distribution",\n    color_continuous_scale=px.colors.sequential.Plasma,\n    mapbox_style="open-street-map"\n)\n\nfig.show(renderer=\'browser\')\n'

#### **Further Handling Missing Values Appropriately**

In [938]:
missing_listing_counts = df_listings_clean.isnull().sum()
missing_listing_counts

OriginalListPrice                0
ListingKey                       0
ListAgentEmail                   0
CloseDate                   461731
ClosePrice                  475198
ListAgentFirstName               0
ListAgentLastName                0
Latitude                         0
Longitude                        0
UnparsedAddress                  0
PropertyType                     0
LivingArea                   43230
ListPrice                        0
DaysOnMarket                     0
ListOfficeName                   0
ListAgentFullName                0
ListingKeyNumeric                0
MLSAreaMajor                 65007
CountyOrParish                   0
MlsStatus                        0
AttachedGarageYN            209037
ParkingTotal                     0
PropertySubType              50789
LotSizeAcres                 60366
SubdivisionName             452735
YearBuilt                    19689
StreetNumberNumeric              0
ListingId                        0
BathroomsTotalIntege

In [939]:
missing_listing_percent = (df_listings_clean.isnull().mean()) * 100
missing_listing_percent

OriginalListPrice            0.000000
ListingKey                   0.000000
ListAgentEmail               0.000000
CloseDate                   67.926091
ClosePrice                  69.907246
ListAgentFirstName           0.000000
ListAgentLastName            0.000000
Latitude                     0.000000
Longitude                    0.000000
UnparsedAddress              0.000000
PropertyType                 0.000000
LivingArea                   6.359644
ListPrice                    0.000000
DaysOnMarket                 0.000000
ListOfficeName               0.000000
ListAgentFullName            0.000000
ListingKeyNumeric            0.000000
MLSAreaMajor                 9.563299
CountyOrParish               0.000000
MlsStatus                    0.000000
AttachedGarageYN            30.751815
ParkingTotal                 0.000000
PropertySubType              7.471663
LotSizeAcres                 8.880553
SubdivisionName             66.602673
YearBuilt                    2.896485
StreetNumber

In [940]:
missing_summary = pd.DataFrame({
    "missing_listing_counts": missing_listing_counts,
    "missing_listing_percent": missing_listing_percent
})

In [941]:
missing_listing_summary = missing_summary.sort_values(by="missing_listing_percent", ascending=False)
print(missing_listing_summary)

                          missing_listing_counts  missing_listing_percent
ClosePrice                                475198                69.907246
CloseDate                                 461731                67.926091
SubdivisionName                           452735                66.602673
PurchaseContractDate                      370125                54.449765
MainLevelBedrooms                         325430                47.874602
HighSchoolDistrict                        240786                35.422468
AssociationFee                            217051                31.930769
AttachedGarageYN                          209037                30.751815
Stories                                   127204                18.713213
GarageSpaces                              100109                14.727218
NewConstructionYN                          84053                12.365190
Levels                                     76496                11.253466
MLSAreaMajor                          

In [942]:
missing_listing_summary[missing_listing_summary['missing_listing_percent'] > 0]

,missing_listing_counts,missing_listing_percent
ClosePrice,475198,69.907246
CloseDate,461731,67.926091
SubdivisionName,452735,66.602673
PurchaseContractDate,370125,54.449765
MainLevelBedrooms,325430,47.874602
HighSchoolDistrict,240786,35.422468
AssociationFee,217051,31.930769
AttachedGarageYN,209037,30.751815
Stories,127204,18.713213
GarageSpaces,100109,14.727218


In [943]:
continuous_columns = df_listings_clean.select_dtypes(include=['int64', 'float64']).columns

In [944]:
#df_listings_clean[continuous_columns].corr(method = "spearman").round(2)

In [945]:
'''corrs =df_listings_clean[continuous_columns].corr(method = "spearman")
mask = np.zeros_like(corrs)
mask[np.triu_indices_from(mask)] = True
sns.heatmap(corrs, cmap='Spectral_r', mask=mask, square=True, vmin=-.4, vmax=.4)
plt.title('Correlation matrix')'''

'corrs =df_listings_clean[continuous_columns].corr(method = "spearman")\nmask = np.zeros_like(corrs)\nmask[np.triu_indices_from(mask)] = True\nsns.heatmap(corrs, cmap=\'Spectral_r\', mask=mask, square=True, vmin=-.4, vmax=.4)\nplt.title(\'Correlation matrix\')'

In [946]:
'''pd.set_option('display.max_rows', None)
pd.set_option('display.max_columns', None)
pd.set_option('display.width', None)
pd.set_option('display.max_colwidth', None)'''

"pd.set_option('display.max_rows', None)\npd.set_option('display.max_columns', None)\npd.set_option('display.width', None)\npd.set_option('display.max_colwidth', None)"

In [947]:
'''corr_matrix = df_listings_clean[continuous_columns].corr(method="spearman").round(2)

moderate_corr = corr_matrix.stack()
moderate_corr = moderate_corr[(moderate_corr.abs() >= 0.3) & (moderate_corr.abs() < 0.7)]

print(moderate_corr)'''

'corr_matrix = df_listings_clean[continuous_columns].corr(method="spearman").round(2)\n\nmoderate_corr = corr_matrix.stack()\nmoderate_corr = moderate_corr[(moderate_corr.abs() >= 0.3) & (moderate_corr.abs() < 0.7)]\n\nprint(moderate_corr)'

- Under 10% missing -> simple statistical imputation method, but for categorical variables, may include a "Missing" category if deemed necessary
- 10-50% missing -> "Decision Zone" - meaning that there are different approaches that can be taken to impute the missing values like a group-based imputation, model-based if justifiable, adding a missing indicator, or doing simple imputation if necessary
- above 50% -> Since at this point, they are probably important features, the missingness in the data could be important and signals themselves. 

Let's start with "Under 10% missing"

In [948]:
under_10 = missing_listing_summary[(missing_listing_summary['missing_listing_percent'] < 10) & (missing_listing_percent != 0)]

C:\Users\mayab\AppData\Local\Temp\ipykernel_30128\2293341584.py:1: UserWarning: Boolean Series key will be reindexed to match DataFrame index.
  under_10 = missing_listing_summary[(missing_listing_summary['missing_listing_percent'] < 10) & (missing_listing_percent != 0)]


In [949]:
under_10

,missing_listing_counts,missing_listing_percent
MLSAreaMajor,65007,9.563299
LotSizeAcres,60366,8.880553
LotSizeSquareFeet,57784,8.500710
LotSizeArea,56830,8.360365
FireplaceYN,56766,8.350950
PropertySubType,50789,7.471663
LivingArea,43230,6.359644
BedroomsTotal,39667,5.835485
YearBuilt,19689,2.896485
BathroomsTotalInteger,15530,2.284647


In [950]:
df_listings_clean['MLSAreaMajor'].describe()

count                614748
unique                 1136
top       699 - Not Defined
freq                  51550
Name: MLSAreaMajor, dtype: object

In [951]:
df_listings_clean['MLSAreaMajor'].value_counts().head()

MLSAreaMajor
699 - Not Defined                     51550
SRCAR - Southwest Riverside County    26686
252 - Riverside                        7239
C14 - Santa Monica                     5484
C42 - Downtown L.A.                    5181
Name: count, dtype: int64

In [952]:
df_listings_clean['MLSAreaMajor'].value_counts(normalize=True).head()

MLSAreaMajor
699 - Not Defined                     0.083855
SRCAR - Southwest Riverside County    0.043410
252 - Riverside                       0.011776
C14 - Santa Monica                    0.008921
C42 - Downtown L.A.                   0.008428
Name: proportion, dtype: float64

In [953]:
df_listings_clean['MLSAreaMajor'] = df_listings_clean['MLSAreaMajor'].fillna("Missing")

In [954]:
df_listings_clean['MLSAreaMajor'].isna().sum()

np.int64(0)

In [955]:
df_listings_clean['LotSizeAcres'].describe()

count    6.193890e+05
mean     6.722773e+02
std      4.706031e+05
min      0.000000e+00
25%      1.200000e-01
50%      1.700000e-01
75%      3.287000e-01
max      3.702600e+08
Name: LotSizeAcres, dtype: float64

In [956]:
df_listings_clean['LotSizeAcres'] = df_listings_clean['LotSizeAcres'].fillna(df_listings_clean['LotSizeAcres'].median())

In [957]:
df_listings_clean['LotSizeAcres'].isna().sum()

np.int64(0)

The mean is greater than the median. The values are skewed to the right. I will be using the median to impute. 

In [958]:
df_listings_clean['LotSizeArea'].describe()

count    6.229250e+05
mean     5.483961e+04
std      2.486892e+06
min      0.000000e+00
25%      4.981000e+03
50%      7.200000e+03
75%      1.250000e+04
max      9.187423e+08
Name: LotSizeArea, dtype: float64

In [959]:
df_listings_clean['LotSizeArea'] = df_listings_clean['LotSizeArea'].fillna(df_listings_clean['LotSizeArea'].median())

In [960]:
df_listings_clean['LotSizeArea'].isna().sum()

np.int64(0)

In [961]:
df_listings_clean['LotSizeArea'].describe()

count    6.797550e+05
mean     5.085676e+04
std      2.380703e+06
min      0.000000e+00
25%      5.150000e+03
50%      7.200000e+03
75%      1.129100e+04
max      9.187423e+08
Name: LotSizeArea, dtype: float64

In [962]:
df_listings_clean['LotSizeSquareFeet'].describe()

count    6.219710e+05
mean     5.397025e+05
std      2.117726e+07
min      0.000000e+00
25%      5.227000e+03
50%      7.405000e+03
75%      1.417650e+04
max      3.472385e+09
Name: LotSizeSquareFeet, dtype: float64

The mean is less than the median. The values are skewed to the left. I will be using the median to impute the missing values. 

In [963]:
df_listings_clean['LotSizeSquareFeet'] = df_listings_clean['LotSizeSquareFeet'].fillna(df_listings_clean['LotSizeSquareFeet'].median())

In [964]:
df_listings_clean['LotSizeSquareFeet'].isna().sum()

np.int64(0)

In [965]:
df_listings_clean['LotSizeSquareFeet'].describe()

count    6.797550e+05
mean     4.944534e+05
std      2.025771e+07
min      0.000000e+00
25%      5.454000e+03
50%      7.405000e+03
75%      1.252900e+04
max      3.472385e+09
Name: LotSizeSquareFeet, dtype: float64

In [966]:
df_listings_clean['FireplaceYN'].value_counts()

FireplaceYN
True     372295
False    250694
Name: count, dtype: int64

In [967]:
df_listings_clean['FireplaceYN'].value_counts(normalize=True)

FireplaceYN
True     0.597595
False    0.402405
Name: proportion, dtype: float64

In [968]:
df_listings_clean['FireplaceYN'] = df_listings_clean['FireplaceYN'].fillna(df_listings_clean['FireplaceYN'].mode()[0])

In [969]:
df_listings_clean['FireplaceYN'].value_counts()

FireplaceYN
True     429061
False    250694
Name: count, dtype: int64

In [970]:
df_listings_clean['FireplaceYN'].value_counts(normalize=True)

FireplaceYN
True     0.631199
False    0.368801
Name: proportion, dtype: float64

In [971]:
df_listings_clean['FireplaceYN'].isna().sum()

np.int64(0)

In [972]:
df_listings_clean['PropertySubType'].value_counts().head()

PropertySubType
SingleFamilyResidence    404416
Condominium              128617
Townhouse                 36597
Apartment                 17416
Duplex                    14438
Name: count, dtype: int64

In [973]:
df_listings_clean['PropertySubType'].value_counts(normalize=True).head()

PropertySubType
SingleFamilyResidence    0.642985
Condominium              0.204490
Townhouse                0.058186
Apartment                0.027690
Duplex                   0.022955
Name: proportion, dtype: float64

In [974]:
df_listings_clean['PropertySubType'] = df_listings_clean['PropertySubType'].fillna(df_listings_clean['PropertySubType'].mode()[0])

In [975]:
df_listings_clean['PropertySubType'].value_counts(normalize=True).head()

PropertySubType
SingleFamilyResidence    0.669660
Condominium              0.189211
Townhouse                0.053839
Apartment                0.025621
Duplex                   0.021240
Name: proportion, dtype: float64

In [976]:
df_listings_clean['LivingArea'].describe()

count    6.365250e+05
mean     3.377681e+03
std      1.139662e+06
min      0.000000e+00
25%      1.178000e+03
50%      1.600000e+03
75%      2.236000e+03
max      9.090909e+08
Name: LivingArea, dtype: float64

The mean is greater than the median, which shows that it is skewing to the right. I will then impute by the median rather than the mean.

In [977]:
df_listings_clean['LivingArea'] = df_listings_clean['LivingArea'].fillna(df_listings_clean['LivingArea'].median())

In [978]:
df_listings_clean['LivingArea'].describe()

count    6.797550e+05
mean     3.264627e+03
std      1.102828e+06
min      0.000000e+00
25%      1.202000e+03
50%      1.600000e+03
75%      2.174000e+03
max      9.090909e+08
Name: LivingArea, dtype: float64

In [979]:
df_listings_clean['LivingArea'].isna().sum()

np.int64(0)

In [980]:
df_listings_clean['BedroomsTotal'].value_counts().head()

BedroomsTotal
3.0    225807
2.0    152451
4.0    145903
5.0     47802
1.0     42389
Name: count, dtype: int64

In [981]:
df_listings_clean['BedroomsTotal'].value_counts(normalize=True).head()

BedroomsTotal
3.0    0.352775
2.0    0.238172
4.0    0.227942
5.0    0.074680
1.0    0.066224
Name: proportion, dtype: float64

Will add a missing indicator for the BedroomsTotal variable to see if the missingness has a real signal moving forward, but will impute based on the median

In [982]:
df_listings_clean['BedroomsTotal_Missing'] = df_listings_clean['BedroomsTotal'].isna().astype(int)

In [983]:
df_listings_clean['BedroomsTotal'] = df_listings_clean['BedroomsTotal'].fillna(df_listings_clean['BedroomsTotal'].median())

In [984]:
df_listings_clean.shape

(679755, 50)

In [985]:
df_listings_clean['BedroomsTotal_Missing'].value_counts()

BedroomsTotal_Missing
0    640088
1     39667
Name: count, dtype: int64

In [986]:
df_listings_clean['BedroomsTotal'].value_counts(normalize=True).head()

BedroomsTotal
3.0    0.390544
2.0    0.224273
4.0    0.214641
5.0    0.070322
1.0    0.062359
Name: proportion, dtype: float64

Will do the same for YearBuilt

In [987]:
df_listings_clean['YearBuilt'].describe()

count    660066.000000
mean       1978.507466
std          28.077942
min        1776.000000
25%        1960.000000
50%        1979.000000
75%        2001.000000
max        2028.000000
Name: YearBuilt, dtype: float64

In [988]:
df_listings_clean['YearBuilt_Missing'] = df_listings_clean['YearBuilt'].isna().astype(int)

In [989]:
df_listings_clean['YearBuilt'] = df_listings_clean['YearBuilt'].fillna(round(df_listings_clean['YearBuilt'].mean()))

In [990]:
(df_listings_clean['YearBuilt'] % 1 == 0).value_counts()

YearBuilt
True    679755
Name: count, dtype: int64

As well as BathroomsTotalInteger

In [991]:
df_listings_clean['BathroomsTotalInteger'].describe()

count    664225.000000
mean          2.525453
std           3.242633
min           0.000000
25%           2.000000
50%           2.000000
75%           3.000000
max        2208.000000
Name: BathroomsTotalInteger, dtype: float64

In [992]:
df_listings_clean['BathroomsTotalInteger'] = df_listings_clean['BathroomsTotalInteger'].fillna(round(df_listings_clean['BathroomsTotalInteger'].mean()))

In [993]:
df_listings_clean['BathroomsTotalInteger'].value_counts(normalize=True).head()

BathroomsTotalInteger
2.0    0.381564
3.0    0.306608
1.0    0.137435
4.0    0.076105
0.0    0.036609
Name: proportion, dtype: float64

Onto the 10-50% missing

In [994]:
btwn_10_50 = missing_listing_summary[(missing_listing_summary['missing_listing_percent'] >= 10) & (missing_listing_percent <= 50)]

C:\Users\mayab\AppData\Local\Temp\ipykernel_30128\3278696572.py:1: UserWarning: Boolean Series key will be reindexed to match DataFrame index.
  btwn_10_50 = missing_listing_summary[(missing_listing_summary['missing_listing_percent'] >= 10) & (missing_listing_percent <= 50)]


In [995]:
btwn_10_50

,missing_listing_counts,missing_listing_percent
MainLevelBedrooms,325430,47.874602
HighSchoolDistrict,240786,35.422468
AssociationFee,217051,31.930769
AttachedGarageYN,209037,30.751815
Stories,127204,18.713213
GarageSpaces,100109,14.727218
NewConstructionYN,84053,12.365190
Levels,76496,11.253466


In [996]:
df_listings_clean['MainLevelBedrooms'].describe()

count    354325.000000
mean          1.933821
std           5.891019
min           0.000000
25%           1.000000
50%           2.000000
75%           3.000000
max        3333.000000
Name: MainLevelBedrooms, dtype: float64

In [998]:
df_listings_clean['MainLevelBedrooms'].value_counts(normalize=True).head()

MainLevelBedrooms
1.0    0.258822
3.0    0.240895
2.0    0.195470
0.0    0.177360
4.0    0.105386
Name: proportion, dtype: float64

In [1000]:
df_listings_clean['MainLevelBedrooms_Missing'] = df_listings_clean['MainLevelBedrooms'].isna().astype(int)

In [1001]:
df_listings_clean['MainLevelBedrooms'] = df_listings_clean['MainLevelBedrooms'].fillna(df_listings_clean['MainLevelBedrooms'].median())

In [1002]:
df_listings_clean['MainLevelBedrooms'].describe()

count    679755.000000
mean          1.965504
std           4.253319
min           0.000000
25%           2.000000
50%           2.000000
75%           2.000000
max        3333.000000
Name: MainLevelBedrooms, dtype: float64

In [1003]:
df_listings_clean['MainLevelBedrooms'].value_counts(normalize=True).head()

MainLevelBedrooms
2.0    0.580636
1.0    0.134912
3.0    0.125567
0.0    0.092449
4.0    0.054933
Name: proportion, dtype: float64

The mean and the median is about the same, so using mean to impute the missing values but also add a missing indicator.

In [883]:
df_listings_clean['MainLevelBedrooms'].isna().sum()

np.int64(0)

In [697]:
df_listings_clean['HighSchoolDistrict'].value_counts().head()

HighSchoolDistrict
Los Angeles Unified          48399
Other                        33035
Capistrano Unified           14611
Irvine Unified               10259
Saddleback Valley Unified     8537
Name: count, dtype: int64

In [698]:
df_listings_clean['HighSchoolDistrict'].value_counts(normalize=True).head()

HighSchoolDistrict
Los Angeles Unified          0.110256
Other                        0.075256
Capistrano Unified           0.033285
Irvine Unified               0.023371
Saddleback Valley Unified    0.019448
Name: proportion, dtype: float64

In [699]:
df_listings_clean['HighSchoolDistrict'] = df_listings_clean['HighSchoolDistrict'].fillna("Missing")

In [700]:
df_listings_clean['HighSchoolDistrict'].value_counts()

HighSchoolDistrict
Missing                                                  240786
Los Angeles Unified                                       48399
Other                                                     33035
Capistrano Unified                                        14611
Irvine Unified                                            10259
Saddleback Valley Unified                                  8537
Long Beach Unified                                         7454
Newport Mesa Unified                                       7344
San Diego Unified                                          6575
Corona-Norco Unified                                       6212
Hemet Unified                                              6212
Riverside Unified                                          6174
William S. Hart Union                                      6077
Huntington Beach Union High                                5892
Palm Springs Unified                                       4726
San Bernardino City U

In [701]:
df_listings_clean['HighSchoolDistrict'].isna().sum()

np.int64(0)

In [702]:
df_listings_clean['AssociationFee'].describe()

count    462704.000000
mean        215.163756
std        2241.751351
min           0.000000
25%           0.000000
50%           0.000000
75%         325.000000
max      968348.000000
Name: AssociationFee, dtype: float64

In [703]:
# Regression Predictions
def impute_lgbm_numeric(df, target):
    df = df.copy()

    # mask missing values
    missing_mask = df[target].isna()

    if missing_mask.sum() == 0:
        return df

    # split data
    train = df[~missing_mask]
    test = df[missing_mask]

    y_train = train[target]
    X_train = train.drop(columns=[target]).select_dtypes(include='number')
    X_test = test.drop(columns=[target]).select_dtypes(include='number')

    # encode categorical columns safely
    for col in X_train.select_dtypes(include=['object', 'category']).columns:
        X_train[col] = X_train[col].astype('category').cat.codes
        X_test[col] = X_test[col].astype('category').cat.codes

    # model
    model = LGBMRegressor(
        n_estimators=200,
        learning_rate=0.05,
        random_state=42,
        verbosity=-1
    )

    model.fit(X_train, y_train)

    preds = model.predict(X_test)

    # constraint: association_fee cannot be negative
    preds = np.clip(preds, a_min=0, a_max=None)

    df.loc[missing_mask, target] = preds

    return df

In [705]:
df_listings_clean = impute_lgbm_numeric(df_listings_clean, "AssociationFee")

In [706]:
df_listings_clean['AssociationFee'].describe()

count    679755.000000
mean        232.897854
std        1876.224280
min           0.000000
25%           0.000000
50%          95.000000
75%         305.873620
max      968348.000000
Name: AssociationFee, dtype: float64

In [707]:
df_listings_clean['AssociationFee'].isna().sum()

np.int64(0)

In [708]:
(df_listings_clean['AssociationFee'] < 0).value_counts()

AssociationFee
False    679755
Name: count, dtype: int64

In [709]:
df_listings_clean['AttachedGarageYN'].value_counts()

AttachedGarageYN
True     362910
False    107808
Name: count, dtype: int64

In [710]:
df_listings_clean['AttachedGarageYN'] = df_listings_clean['AttachedGarageYN'].fillna("Missing")

In [711]:
df_listings_clean['AttachedGarageYN'].value_counts()

AttachedGarageYN
True       362910
Missing    209037
False      107808
Name: count, dtype: int64

In [712]:
df_listings_clean['Stories'].describe()

count    552551.000000
mean          1.363941
std           0.481132
min           1.000000
25%           1.000000
50%           1.000000
75%           2.000000
max           2.000000
Name: Stories, dtype: float64

In [713]:
df_listings_clean['Stories'].value_counts()

Stories
1.0    351455
2.0    201096
Name: count, dtype: int64

In [714]:
df_listings_clean['Stories'].value_counts(normalize=True)

Stories
1.0    0.636059
2.0    0.363941
Name: proportion, dtype: float64

In [715]:
df_listings_clean['Stories_Missing'] = df_listings_clean['Stories'].isna().astype(int)

In [716]:
df_listings_clean['Stories'] = df_listings_clean['Stories'].fillna(df_listings_clean['Stories'].median())

In [717]:
df_listings_clean['Stories'].value_counts()

Stories
1.0    478659
2.0    201096
Name: count, dtype: int64

In [718]:
df_listings_clean['Stories'].isna().sum()

np.int64(0)

In [719]:
df_listings_clean['GarageSpaces'].value_counts().head()

GarageSpaces
2.0    327434
0.0    103010
1.0     70882
3.0     61391
4.0     10167
Name: count, dtype: int64

In [720]:
df_listings_clean['GarageSpaces'].value_counts(normalize=True).head()

GarageSpaces
2.0    0.564886
0.0    0.177712
1.0    0.122285
3.0    0.105911
4.0    0.017540
Name: proportion, dtype: float64

In [721]:
df_listings_clean['GarageSpaces'].unique()

array([  2.  ,    nan,   3.  ,   4.  ,   0.  ,   1.  ,   6.  ,   7.  ,
         2.5 ,   3.5 ,   8.  ,   5.  ,  13.  ,   9.  ,  14.  ,  25.  ,
        10.  , 400.  ,  16.  ,  12.  ,  52.  ,   1.5 ,  26.  , 200.  ,
        18.  ,   4.5 ,  22.  , 500.  ,  20.  ,  30.  ,  11.  ,  21.  ,
        34.  ,  15.  ,  35.  ,  80.  ,  49.  ,  40.  ,  24.  ,  54.  ,
        50.  ,  44.  , 300.  ,  60.  ,  17.  ,  19.  ,  48.  ,  83.  ,
        33.  , 600.  ,  28.  ,  23.  ,  72.  ,   6.5 ,  31.  ,  27.  ,
        36.  ,  32.  ,   8.5 ,   2.25,  45.  , 100.  ,  78.  ,  38.  ,
        46.  ,  75.  ,  41.  ,  51.  ,  29.  ,  55.  ,  64.  ,  86.  ,
        88.  ,  96.  ,  37.  ])

In [722]:
df_listings_clean['GarageSpaces_Missing'] = df_listings_clean['GarageSpaces'].isna().astype(int)

In [723]:
df_listings_clean['GarageSpaces'] = df_listings_clean['GarageSpaces'].fillna(df_listings_clean['GarageSpaces'].median())

In [724]:
df_listings_clean['GarageSpaces'].value_counts().head()

GarageSpaces
2.0    427543
0.0    103010
1.0     70882
3.0     61391
4.0     10167
Name: count, dtype: int64

In [725]:
df_listings_clean['GarageSpaces'].value_counts(normalize=True).head()

GarageSpaces
2.0    0.628966
0.0    0.151540
1.0    0.104276
3.0    0.090313
4.0    0.014957
Name: proportion, dtype: float64

In [726]:
df_listings_clean['GarageSpaces'].isna().sum()

np.int64(0)

In [727]:
df_listings_clean['NewConstructionYN'].value_counts()

NewConstructionYN
False    567956
True      27746
Name: count, dtype: int64

In [728]:
df_listings_clean['NewConstructionYN'].value_counts(normalize=True)

NewConstructionYN
False    0.953423
True     0.046577
Name: proportion, dtype: float64

In [729]:
df_listings_clean['NewConstructionYN'] = df_listings_clean['NewConstructionYN'].fillna(df_listings_clean['NewConstructionYN'].mode()[0])

In [730]:
df_listings_clean['NewConstructionYN'].value_counts(normalize=True)

NewConstructionYN
False    0.959182
True     0.040818
Name: proportion, dtype: float64

In [731]:
df_listings_clean['NewConstructionYN'].isna().sum()

np.int64(0)

In [732]:
df_listings_clean['Levels'].value_counts().head()

Levels
One            347099
Two            199429
ThreeOrMore     31921
MultiSplit      18102
One,Two          2223
Name: count, dtype: int64

In [733]:
df_listings_clean['Levels'].value_counts(normalize=True).head()

Levels
One            0.575373
Two            0.330586
ThreeOrMore    0.052914
MultiSplit     0.030007
One,Two        0.003685
Name: proportion, dtype: float64

In [734]:
df_listings_clean['Levels'].unique()

<StringArray>
[                           'One',                            'Two',
                    'ThreeOrMore',                     'MultiSplit',
                              nan,                        'One,Two',
                'Two,ThreeOrMore',     'Two,ThreeOrMore,MultiSplit',
             'One,Two,MultiSplit',         'ThreeOrMore,MultiSplit',
                 'Two,MultiSplit',                'One,ThreeOrMore',
                 'One,MultiSplit',            'One,Two,ThreeOrMore',
     'One,ThreeOrMore,MultiSplit', 'One,Two,ThreeOrMore,MultiSplit',
                        'Two,One',                 'MultiSplit,One',
             'Two,MultiSplit,One',                'ThreeOrMore,One']
Length: 20, dtype: str

In [735]:
df_listings_clean['Levels'] = df_listings_clean['Levels'].fillna("Missing")

In [736]:
df_listings_clean['Levels'].value_counts(normalize=True).head()

Levels
One            0.510624
Two            0.293384
Missing        0.112535
ThreeOrMore    0.046960
MultiSplit     0.026630
Name: proportion, dtype: float64

In [737]:
df_listings_clean['Levels'].isna().sum()

np.int64(0)

Over 50% missing values

In [738]:
over_50 = missing_listing_summary[missing_listing_summary['missing_listing_percent'] >= 50]

In [739]:
over_50

,missing_listing_counts,missing_listing_percent
ClosePrice,475198,69.907246
CloseDate,461731,67.926091
SubdivisionName,452735,66.602673
PurchaseContractDate,370125,54.449765


In [740]:
df_listings_clean['ClosePrice'].describe()

count    2.045570e+05
mean     7.510010e+05
std      3.650523e+06
min      0.000000e+00
25%      4.800000e+03
50%      5.500000e+05
75%      9.750000e+05
max      8.200000e+08
Name: ClosePrice, dtype: float64

In [741]:
df_listings_clean["ClosePrice_Missing"] = df_listings_clean["ClosePrice"].isna().astype(int)

In [742]:
df_listings_clean['ClosePrice_Missing'].value_counts()

ClosePrice_Missing
1    475198
0    204557
Name: count, dtype: int64

In [743]:
df_listings_clean['ClosePrice'].describe()

count    2.045570e+05
mean     7.510010e+05
std      3.650523e+06
min      0.000000e+00
25%      4.800000e+03
50%      5.500000e+05
75%      9.750000e+05
max      8.200000e+08
Name: ClosePrice, dtype: float64

In [744]:
df_listings_clean['CloseDate'].value_counts()

CloseDate
2025-10-31    1289
2025-11-21    1272
2026-02-13    1244
2025-08-29    1155
2025-09-30    1150
2025-10-10    1150
2025-11-14    1144
2025-10-24    1130
2025-11-26    1126
2025-12-05    1118
2025-12-19    1112
2025-12-01    1111
2026-01-30    1107
2025-10-15    1098
2025-11-07    1063
2025-12-12    1049
2025-10-17    1041
2025-10-01    1038
2024-03-29     997
2025-09-29     984
2025-11-25     984
2024-04-30     983
2025-09-19     975
2025-09-26     974
2025-09-12     974
2026-01-16     968
2025-12-15     964
2025-12-30     955
2025-08-15     942
2025-11-24     942
2025-10-03     940
2025-12-23     934
2025-11-10     929
2025-09-05     916
2025-11-13     905
2025-10-27     905
2025-09-15     896
2025-11-17     885
2025-12-08     885
2025-08-22     885
2025-10-30     884
2025-10-14     882
2026-02-17     880
2026-02-06     879
2024-04-26     869
2024-05-01     865
2025-08-01     862
2025-11-18     845
2026-02-18     843
2024-04-12     840
2025-12-22     835
2025-12-31     833
20

In [745]:
missing_close_date = df_listings_clean[df_listings_clean['CloseDate'].isna()]

In [746]:
missing_close_date.shape

(461731, 56)

In [747]:
missing_close_date['PurchaseContractDate'].isna().sum()

np.int64(364044)

In [748]:
missing_close_date['PurchaseContractDate'].isna().sum() / missing_close_date.shape[0]

np.float64(0.7884330919951227)

About 79% of missing close price entries have missing purchase contract date

This could signal that the missingness in close price is associated with another missing variable which is purchase contract date. So, the data could not be completely missing at random. 

In [749]:
pd.crosstab(
    df_listings_clean['CloseDate'].isna(),
    df_listings_clean['PurchaseContractDate'].isna(),
    normalize='index'
)

PurchaseContractDate,False,True
CloseDate,,
False,0.972109,0.027891
True,0.211567,0.788433


In [750]:
print(df_listings_clean['PurchaseContractDate'].dtype)

datetime64[us]


In [751]:
print(df_listings_clean['CloseDate'].dtype)

datetime64[us]


In [752]:
df_listings_clean['No_Purchase_Date'] = df_listings_clean['PurchaseContractDate'].isna()
df_listings_clean['No_Close_Date'] = df_listings_clean['CloseDate'].isna()

In [753]:
df_listings_clean['PurchaseContractDate'].isna().sum()

np.int64(370125)

In [754]:
df_listings_clean['CloseDate'].value_counts(dropna=False).head()

CloseDate
NaT           461731
2025-10-31      1289
2025-11-21      1272
2026-02-13      1244
2025-08-29      1155
Name: count, dtype: int64

In [755]:
df_listings_clean['PurchaseContractDate'].value_counts(dropna=False).head()

PurchaseContractDate
NaT           370125
2025-10-01       958
2026-02-02       943
2026-01-26       942
2026-02-09       899
Name: count, dtype: int64

In [756]:
df_listings_clean['No_Purchase_Date'].value_counts()

No_Purchase_Date
True     370125
False    309630
Name: count, dtype: int64

In [757]:
df_listings_clean['No_Close_Date'].value_counts()

No_Close_Date
True     461731
False    218024
Name: count, dtype: int64

Keep the date variables with NaT to show those events were not observed or did not occur - I also created corresponding variables to measure this

In [758]:
df_listings_clean['SubdivisionName'].value_counts().head()

SubdivisionName
Not Applicable-1      7086
Other                 6243
Not Applicable        3795
Leisure World (LW)    2649
San Diego             2430
Name: count, dtype: int64

In [759]:
df_listings_clean['SubdivisionName'] = df_listings_clean['SubdivisionName'].fillna("Missing")

In [760]:
df_listings_clean['SubdivisionName'].value_counts().head()

SubdivisionName
Missing               452735
Not Applicable-1        7086
Other                   6243
Not Applicable          3795
Leisure World (LW)      2649
Name: count, dtype: int64

In [761]:
df_listings_clean.isna().sum().sum()

np.int64(1307054)

In [762]:
df_listings_clean[['ClosePrice', 'CloseDate', 'PurchaseContractDate']].isna().sum()

ClosePrice              475198
CloseDate               461731
PurchaseContractDate    370125
dtype: int64

In [763]:
df_listings_clean.shape

(679755, 58)

In [764]:
df_listings_clean.columns

Index(['OriginalListPrice', 'ListingKey', 'ListAgentEmail', 'CloseDate',
       'ClosePrice', 'ListAgentFirstName', 'ListAgentLastName', 'Latitude',
       'Longitude', 'UnparsedAddress', 'PropertyType', 'LivingArea',
       'ListPrice', 'DaysOnMarket', 'ListOfficeName', 'ListAgentFullName',
       'ListingKeyNumeric', 'MLSAreaMajor', 'CountyOrParish', 'MlsStatus',
       'AttachedGarageYN', 'ParkingTotal', 'PropertySubType', 'LotSizeAcres',
       'SubdivisionName', 'YearBuilt', 'StreetNumberNumeric', 'ListingId',
       'BathroomsTotalInteger', 'City', 'BedroomsTotal',
       'ContractStatusChangeDate', 'PurchaseContractDate',
       'ListingContractDate', 'StateOrProvince', 'FireplaceYN', 'Stories',
       'Levels', 'LotSizeArea', 'MainLevelBedrooms', 'NewConstructionYN',
       'GarageSpaces', 'HighSchoolDistrict', 'PostalCode', 'AssociationFee',
       'LotSizeSquareFeet', 'Year', 'year_month', 'rate_30yr_fixed',
       'BedroomsTotal_Missing', 'YearBuilt_Missing',
       'Bathroo

In [1004]:
df_listings_clean.shape

(679755, 52)

In [1005]:
df_listings_clean.isna().sum().sum()

np.int64(2814525)

In [1007]:
df_listings_clean.isna().sum().sum() / df_listings_clean.notna().sum().sum()

np.float64(0.08651363004063446)

In [1008]:
df_listings_clean.columns

Index(['OriginalListPrice', 'ListingKey', 'ListAgentEmail', 'CloseDate',
       'ClosePrice', 'ListAgentFirstName', 'ListAgentLastName', 'Latitude',
       'Longitude', 'UnparsedAddress', 'PropertyType', 'LivingArea',
       'ListPrice', 'DaysOnMarket', 'ListOfficeName', 'ListAgentFullName',
       'ListingKeyNumeric', 'MLSAreaMajor', 'CountyOrParish', 'MlsStatus',
       'AttachedGarageYN', 'ParkingTotal', 'PropertySubType', 'LotSizeAcres',
       'SubdivisionName', 'YearBuilt', 'StreetNumberNumeric', 'ListingId',
       'BathroomsTotalInteger', 'City', 'BedroomsTotal',
       'ContractStatusChangeDate', 'PurchaseContractDate',
       'ListingContractDate', 'StateOrProvince', 'FireplaceYN', 'Stories',
       'Levels', 'LotSizeArea', 'MainLevelBedrooms', 'NewConstructionYN',
       'GarageSpaces', 'HighSchoolDistrict', 'PostalCode', 'AssociationFee',
       'LotSizeSquareFeet', 'Year', 'year_month', 'rate_30yr_fixed',
       'BedroomsTotal_Missing', 'YearBuilt_Missing',
       'MainLev

In [765]:
#df_listings_clean.to_csv("../data/Listing/listings_cleaned.csv", index=False)